pkg/endpoint/local_ip.go

```diff
@@ -53,12 +53,25 @@ func getLocalIPFromRoutes(family k8snet.IPFamily) (string, error) {
 }
 
 func GetLocalIPForDestination(dst string, family k8snet.IPFamily) string {
-	conn, err := Dial("udp"+string(family), "["+dst+"]:53")
+	conn, err := Dial("udp"+string(family), net.JoinHostPort(dst, "53"))
 	if err == nil {
 		defer conn.Close()
 		localAddr := conn.LocalAddr().(*net.UDPAddr)
 
-		return localAddr.IP.String()
+		if family == k8snet.IPv4 {
+			return localAddr.IP.String()
+		}
+
+		if !isValidGlobalIPv6(localAddr.IP) {
+			logger.Warningf("IP %q returned from Dial isn't usable - trying to find another IP from the same interface", localAddr.IP)
+			// Try to find a valid global-scope IP from the same interface
+			ifaceIP, err := getValidGlobalIPv6FromSameInterface(localAddr.IP)
+			if err == nil {
+				return ifaceIP
+			} else {
+				logger.Warningf("Failed to retrieve valid IPv6 address from interface for %q", localAddr.IP)
+			}
+		}
```

Comment:
I think it needs to return if IPv6 is valid:

```suggestion
		if family == k8snet.IPv4 || isValidGlobalIPv6(localAddr.IP) {
			return localAddr.IP.String()
		}

		logger.Warningf("IP %q returned from Dial isn't usable - trying to find another IP from the same interface", localAddr.IP)
		// Try to find a valid global-scope IP from the same interface
		...
```